File: backends/dotnet/Controllers/Todos.controller.cs
using Dotnet.Contexts;
using Dotnet.Models;
using Microsoft.AspNetCore.Mvc;

namespace Dotnet.Controllers
{
  [ApiController]
  [Route("[controller]")]
  public class TodosController(DatabaseContext databaseContext) : ControllerBase
  {
    private readonly DatabaseContext _databaseContext = databaseContext;

    private Todo? Find(string id) => _databaseContext.Todos.Find(id);

    [HttpGet]
    public IActionResult GetAll() => Ok(_databaseContext.Todos.ToList());

    [HttpPost]
    public IActionResult Create([FromBody] Todo todo)
    {
      try
      {
        if (string.IsNullOrEmpty(todo.Title))
          throw new Exception("Title cannot be null or empty");

        _databaseContext.Add(todo);
        _databaseContext.SaveChanges();

        return CreatedAtAction(nameof(GetAll), new { todo.Id }, todo);
      }
      catch (Exception e)
      {
        return BadRequest(new { error = e.Message });
      }
    }

    [HttpPut("{id}")]
    public IActionResult Update([FromRoute] string id, [FromBody] Todo todo)
    {
      Todo? entity = Find(id);

      if (entity is null) return NotFound();

      entity.Title = todo.Title;
      entity.Completed = todo.Completed;

      _databaseContext.SaveChanges();

      return Ok();
    }

    [HttpDelete("{id}")]
    public IActionResult Delete([FromRoute] string id)
    {
      Todo? entity = Find(id);

      if (entity is null) return NotFound();

      _databaseContext.Remove(entity);
      _databaseContext.SaveChanges();

      return NoContent();
    }
  }
}

File: backends/node/controllers/todos.controller.js
const express = require('express');
const router = express.Router();
const context = require('../contexts/context');

router.get('/', async (req, res) => {
  const todos = await context.todos.findMany();
  res.json(todos);
});

router.post('/', async (req, res) => {
  const { title, completed } = req.body;

  if (!title || title.trim() === '')
    return res.status(400).json({ error: 'Title cannot be null or empty' });

  const newTodo = await context.todos.create({
    data: { Title: title, Completed: !!completed },
  });

  res.status(201).json(newTodo);
});

router.put('/:id', async (req, res) => {
  const { id } = req.params;
  const { title, completed } = req.body;
  const entity = await context.todos.findUnique({ where: { Id: id } });

  if (!entity)
    return res.status(404).send();

  const updated = await context.todos.update({
    where: { Id: id },
    data: { Title: title, Completed: completed },
  });

  res.status(200).json(updated);
});

router.delete('/:id', async (req, res) => {
  const { id } = req.params;
  const entity = await context.todos.findUnique({ where: { Id: id } });

  if (!entity)
    return res.status(404).send();

  await context.todos.delete({ where: { Id: id } });
  res.status(204).send();
});

module.exports = router;